File: Formatter.java
import java.io.BufferedReader;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
//import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Formatter {

    private static BufferedReader br;

    private static final ArrayList<String> list_of_bad_words = new ArrayList<String>() {
        static final long serialVersionUID = 1L;
        {
            try {
                br = new BufferedReader(new FileReader(".\\resources\\bad_words.txt"));
                String line;
                while ((line = br.readLine()) != null) {
                    if (!line.contains("#")) {
                        add(line);
                    }
                }
                br.close();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    };

    private static final ArrayList<String> list_of_correct_words = new ArrayList<String>() {
        static final long serialVersionUID = 1L;
        {
            try {
                br = new BufferedReader(new FileReader(".\\resources\\good_words.txt"));
                String line;
                while ((line = br.readLine()) != null) {
                    if (!line.contains("#")) {
                        add(line);
                    }
                }
                br.close();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    };

    private static final ArrayList<String> list_of_function_names = new ArrayList<String>() {
        static final long serialVersionUID = 1L;
        {
            try {
                br = new BufferedReader(new FileReader(".\\resources\\function_names.txt"));
                String line;
                while ((line = br.readLine()) != null) {
                    if (!line.contains("#")) {
                        add(line);
                    }
                }
                br.close();
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    };

    public static void main(String[] args) {
        // Copied from:
        // https://stackoverflow.com/questions/3154488/how-do-i-iterate-through-the-files-in-a-directory-in-java
        File[] directoryListing = new File(args[0]).listFiles();
        long startTime = 0L;
        if (directoryListing != null) {
            // Copied from:
            // https://stackoverflow.com/questions/3382954/measure-execution-time-for-a-java-method
            startTime = System.currentTimeMillis();
            for (File child : directoryListing) {
                if (child.isDirectory()) {
                    for (File file : child.listFiles()) {
                        if (!getExtension(file).equals(".rdl")) {
                            processFile(file);
                        }
                    }
                }
                processFile(child);
            }
        }
        System.out.println("\nExecution-Time: " + ((System.currentTimeMillis() - startTime) / 1000) + " seconds");
    }

    // Copied from:
    // https://stackoverflow.com/questions/3571223/how-do-i-get-the-file-extension-of-a-file-in-java
    private static String getExtension(File file) {
        String extension = "";
        String fileName = file.getName();
        int i = fileName.lastIndexOf('.');
        int p = Math.max(fileName.lastIndexOf('/'), fileName.lastIndexOf('\\'));
        if (i > p) {
            extension = fileName.substring(i + 1);
        }
        return extension;
    }

    private static void processFile(File file) {
        boolean bad_word_found = false;
        try {
            br = new BufferedReader(new FileReader(file));
            StringBuffer inputBuffer = new StringBuffer();
            String line;
            while ((line = br.readLine()) != null) {
                inputBuffer.append(line);
                inputBuffer.append('\n');
            }
            if (inputBuffer.length() > 0) {
                inputBuffer.deleteCharAt(inputBuffer.length() - 1);
            }
            br.close();
            String inputStr = inputBuffer.toString();
            for (String bad_word : list_of_bad_words) {
                if (inputStr.contains(bad_word)) {
                    for (String good_word : list_of_correct_words) {
                        if (bad_word.equalsIgnoreCase(good_word)) {
                            bad_word_found = true;
                            inputStr = inputStr.replace(bad_word, good_word);
                        }
                    }
                }
            }
            for (String function_name : list_of_function_names) {
                Pattern pattern = Pattern.compile(function_name + ";");
                Matcher matcher = pattern.matcher(inputStr);
                if (matcher.find()) {
                    for (String good_word : list_of_correct_words) {
                        if (function_name.equalsIgnoreCase(good_word)) {
                            inputStr = inputStr.replace(function_name + ";", good_word + "();");
                        }
                    }
                }
            }
            // Copied from:
            // https://stackoverflow.com/questions/632204/java-string-replace-using-regular-expressions
            // https://stackoverflow.com/questions/16866077/regex-using-java-string-replaceall
            inputStr = Pattern.compile("\\)\n\\s*\\{\n\\s*\\}").matcher(inputStr).replaceAll(") { }");
            if (bad_word_found) {
                FileOutputStream fileOut = new FileOutputStream(file.getAbsolutePath());
                fileOut.write(inputStr.getBytes());
                fileOut.close();
            }
        } catch (IOException e) {
            return;
        }
    }
}

File: Formatter.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <list>
#include <windows.h>
#include <sstream>
#include <regex>
#include <chrono>
#include <process.h>
#include <thread>

#ifdef WINDOWS
#include <direct.h>
#define GetCurrentDir _getcwd
#else
#include <unistd.h>
#define GetCurrentDir getcwd
#endif

using namespace std;

list<string> readDict(string fileName)
{
    // Copied from:
    // https://gist.github.com/stevedoyle/1319089
    list<string> temp;
    string line;
    ifstream myfile(fileName);
    if (myfile.is_open())
    {
        while (myfile.good())
        {
            getline(myfile, line);
            if (!(line.rfind("#", 0) == 0))
            {
                temp.push_back(line);
            }
        }
        myfile.close();
    }
    else
    {
        printf("%s is not open.", fileName.c_str());
    }
    return temp;
}

string get_current_dir()
{
    char buff[FILENAME_MAX]; // create string buffer to hold path
    GetCurrentDir(buff, FILENAME_MAX);
    string current_working_dir(buff);
    return current_working_dir;
}

const string directory = get_current_dir();
list<string> bad_words = readDict(directory + "\\resources\\bad_words.txt");
list<string> good_words = readDict(directory + "\\resources\\good_words.txt");
list<string> function_names = readDict(directory + "\\resources\\function_names.txt");
bool badWordFound = false;

void replaceAll(string &subject, string search, string replace)
{
    // Copied from:
    // https://thispointer.com/find-and-replace-all-occurrences-of-a-sub-string-in-c/
    size_t pos = subject.find(search);
    while (pos != string::npos)
    {
        subject.replace(pos, search.size(), replace);
        pos = subject.find(search, pos + replace.size());
    }
}

string process(string temp)
{
    string next_help = "NEXT = 0";
    string next_help2 = "Next = 0";
    string temp_for_search = temp;
    for (string bad_word : bad_words)
    {
        if (temp_for_search.find(bad_word) != string::npos)
        {
            for (string good_word : good_words)
            {
                string good_word_for_check = good_word;
                transform(bad_word.begin(), bad_word.end(), bad_word.begin(), ::toupper);
                transform(good_word_for_check.begin(), good_word_for_check.end(), good_word_for_check.begin(), ::toupper);
                if (bad_word == good_word_for_check)
                {
                    badWordFound = true;
                    replaceAll(temp, bad_word, good_word);
                    break;
                }
            }
        }
    }
    for (string function_name : function_names)
    {
        string search_function = function_name + ";";
        string search_function2 = function_name + " then";
        if (temp.find(search_function.c_str()) != string::npos)
        {
            for (string good_word : good_words)
            {
                if (function_name == good_word)
                {
                    badWordFound = true;
                    string new_function = good_word + "();";
                    if (function_name == next_help.c_str() || function_name == next_help2.c_str())
                    {
                        new_function = "Next() = 0;";
                    }
                    replaceAll(temp, search_function, new_function);
                }
            }
        }
        else if (temp.find(search_function2.c_str()) != string::npos)
        {
            for (string good_word : good_words)
            {
                if (function_name == good_word)
                {
                    badWordFound = true;
                    replaceAll(temp, search_function2, good_word + "() then");
                }
            }
        }
    }
    return temp;
}

void processFile(string fileName)
{
    string line;
    ifstream myfile(fileName);
    stringstream os;
    if (myfile.is_open())
    {
        string line;
        while (getline(myfile, line))
        {
            os << line << endl;
        }
    }
    else
    {
        printf("%s is not open.", fileName.c_str());
    }
    string temp = os.str();
    temp = process(temp.substr(0, temp.size() - 1));
    string temp_for_check = temp;
    try
    {
        // Copied from:
        // https://stackoverflow.com/questions/11508798/conditionally-replace-regex-matches-in-string
        temp = regex_replace(temp, regex("\\)\n\\s*\\{\n\\s*\\}"), string(") { }"));
        regex bracket("\\}\n*");
        sregex_iterator iter(temp.begin(), temp.end(), bracket);
        sregex_iterator end;
        int pos;
        while (iter != end)
        {
            pos = iter->position();
            ++iter;
        }
        temp = temp.substr(0, pos + 1);
    }
    catch (regex_error &e)
    {
        // do nothing
    }
    if (badWordFound || temp != temp_for_check)
    {
        ofstream write;
        write.open(fileName, ios::out | ios::binary);
        write << temp;
        write.close();
    }
}

bool hasEnding(string const &fullString, string const &ending)
{
    if (fullString.length() >= ending.length())
    {
        return (0 == fullString.compare(fullString.length() - ending.length(), ending.length(), ending));
    }
    else
    {
        return false;
    }
}

void processFolder(string folder)
{
    // Copied from:
    // https://stackoverflow.com/questions/67273/how-do-you-iterate-through-every-file-directory-recursively-in-standard-c
    string process_path;
    string search_path;
    bool singleFile = hasEnding(folder, ".al");
    if (singleFile)
    {
        process_path = folder;
        search_path = folder;
    }
    else
    {
        process_path = folder + "\\";
        search_path = folder + "\\*";
    }
    WIN32_FIND_DATAA fd;
    HANDLE hFind = ::FindFirstFileA(search_path.c_str(), &fd);
    if (hFind != INVALID_HANDLE_VALUE)
    {
        do
        {
            if (!(fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY))
            {
                string temp;
                if (singleFile)
                {
                    temp = process_path;
                }
                else
                {
                    string ws(fd.cFileName);
                    string s(ws.begin(), ws.end());
                    temp = process_path + s;
                }
                processFile(temp.c_str());
            }
            else if ((fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY))
            {
                std::string temp;
                std::string ws(fd.cFileName);
                std::string s(ws.begin(), ws.end());
                temp = search_path.substr(0, search_path.length() - 1) + s;
                if ((ws != ".") && (ws != ".."))
                {
                    processFolder(temp);
                }
            }
        } while (::FindNextFileA(hFind, &fd));
        ::FindClose(hFind);
    }
}

int main(int argc, char const *argv[])
{
    cout << "Formatting " << argv[1] << endl;
    auto t1 = chrono::high_resolution_clock::now();
    processFolder(argv[1]);
    auto t2 = chrono::high_resolution_clock::now();
    auto duration = chrono::duration_cast<chrono::seconds>(t2 - t1).count();
    if (duration == 0)
    {
        auto duration = chrono::duration_cast<chrono::milliseconds>(t2 - t1).count();
        cout << "\nExecution-Time: " << duration << " milliseconds.\n"
             << endl;
    }
    else
    {
        cout << "\nExecution-Time: " << duration << " seconds.\n"
             << endl;
    }
    chrono::milliseconds dura(500);
    this_thread::sleep_for(dura);
    return 0;
}